processing/src/main/java/io/druid/query/groupby/orderby/DefaultLimitSpec.java

```diff
@@ -320,6 +343,78 @@ public int hashCode()
     }
   }
 
+  public class GroupTopNFunction implements Function<Sequence<Row>, Sequence<Row>>
+  {
+    private final Ordering<Row> ordering;
+    private final int limit;
+    private final QueryGranularity granularity;
+
+    public GroupTopNFunction(Ordering<Row> ordering, QueryGranularity granularity, int limit)
+    {
+      this.ordering = ordering;
+      this.granularity = granularity;
+      this.limit = limit;
+    }
+
+    @Override
+    public Sequence<Row> apply(
+        Sequence<Row> input
+    )
+    {
+      Map<Long, MinMaxPriorityQueue<Row>> grouped = Maps.<Long, MinMaxPriorityQueue<Row>>newTreeMap();
+      ArrayList<Row> rows = Sequences.toList(input, Lists.<Row>newArrayList());
```

Comment:
this means whole result set will be materialized in-memory on the heap potentially doubling the memory requirement for processing the query.
same is true for line 379
this probably needs to be highlighted in documentation.